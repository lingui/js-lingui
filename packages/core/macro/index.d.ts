import type { I18n, MessageDescriptor } from "@lingui/core"

export type ChoiceOptions = {
  /** Offset of value when calculating plural forms */
  offset?: number
  zero?: string
  one?: string
  two?: string
  few?: string
  many?: string

  /** Catch-all option */
  other: string
  /** Exact match form, corresponds to =N rule */
  [digit: `${number}`]: string
}

type MacroMessageDescriptor = (
  | {
      id: string
      message?: string
    }
  | {
      id?: string
      message: string
    }
) & {
  comment?: string
  context?: string
}

/**
 * Translates a message descriptor
 *
 * @example
 * ```
 * import { t } from "@lingui/core/macro";
 * const message = t({
 *   id: "msg.hello",
 *   comment: "Greetings at the homepage",
 *   message: `Hello ${name}`,
 * });
 * ```
 *
 * @example
 * ```
 * import { t } from "@lingui/core/macro";
 * const message = t({
 *   id: "msg.plural",
 *   message: plural(value, { one: "...", other: "..." }),
 * });
 * ```
 *
 * @param descriptor The message descriptor to translate
 */
export function t(descriptor: MacroMessageDescriptor): string

/**
 * Translates a template string using the global I18n instance
 *
 * @example
 * ```
 * import { t } from "@lingui/core/macro";
 * const message = t`Hello ${name}`;
 * ```
 */
export function t(
  literals: TemplateStringsArray,
  ...placeholders: any[]
): string

/**
 * Translates a template string or message descriptor using a given I18n instance
 *
 * @example
 * ```
 * import { t } from "@lingui/core/macro";
 * import { I18n } from "@lingui/core";
 * const i18n = new I18n({
 *   locale: "nl",
 *   messages: { "Hello {0}": "Hallo {0}" },
 * });
 * const message = t(i18n)`Hello ${name}`;
 * ```
 *
 * @example
 * ```
 * import { t } from "@lingui/core/macro";
 * import { I18n } from "@lingui/core";
 * const i18n = new I18n({
 *   locale: "nl",
 *   messages: { "Hello {0}": "Hallo {0}" },
 * });
 * const message = t(i18n)({ message: `Hello ${name}` });
 * ```
 */
export function t(i18n: I18n): {
  (literals: TemplateStringsArray, ...placeholders: any[]): string
  (descriptor: MacroMessageDescriptor): string
}

/**
 * Pluralize a message
 *
 * @example
 * ```
 * import { plural } from "@lingui/core/macro";
 * const message = plural(count, {
 *   one: "# Book",
 *   other: "# Books",
 * });
 * ```
 *
 * @param value Determines the plural form
 * @param options Object with available plural forms
 */
export function plural(value: number | string, options: ChoiceOptions): string

/**
 * Pluralize a message using ordinal forms
 *
 * Similar to `plural` but instead of using cardinal plural forms,
 * it uses ordinal forms.
 *
 * @example
 * ```
 * import { selectOrdinal } from "@lingui/core/macro";
 * const message = selectOrdinal(count, {
 *    one: "#st",
 *    two: "#nd",
 *    few: "#rd",
 *    other: "#th",
 * });
 * ```
 *
 * @param value Determines the plural form
 * @param options Object with available plural forms
 */
export function selectOrdinal(
  value: number | string,
  options: ChoiceOptions
): string

type SelectOptions = {
  /** Catch-all option */
  other: string
  [matches: string]: string
}

/**
 * Selects a translation based on a value
 *
 * Select works like a switch statement. It will
 * select one of the forms in `options` object which
 * key matches exactly `value`.
 *
 * @example
 * ```
 * import { select } from "@lingui/core/macro";
 * const message = select(gender, {
 *    male: "he",
 *    female: "she",
 *    other: "they",
 * });
 * ```
 *
 * @param value The key of choices to use
 * @param choices
 */
export function select(value: string, choices: SelectOptions): string

/**
 * Define a message for later use
 *
 * `defineMessage` can be used to add comments for translators,
 * or to override the message ID.
 *
 * @example
 * ```
 * import { defineMessage } from "@lingui/core/macro";
 * const message = defineMessage({
 *    comment: "Greetings on the welcome page",
 *    message: `Welcome, ${name}!`,
 * });
 * ```
 *
 * @param descriptor The message descriptor
 */
export function defineMessage(
  descriptor: MacroMessageDescriptor
): MessageDescriptor

/**
 * Define a message for later use
 *
 * @example
 * ```
 * import { defineMessage, msg } from "@lingui/core/macro";
 * const message = defineMessage`Hello ${name}`;
 *
 * // or using shorter version
 * const message = msg`Hello ${name}`;
 * ```
 */
export function defineMessage(
  literals: TemplateStringsArray,
  ...placeholders: any[]
): MessageDescriptor

/**
 * Define a message for later use
 * Alias for {@see defineMessage}
 */
export const msg: typeof defineMessage
