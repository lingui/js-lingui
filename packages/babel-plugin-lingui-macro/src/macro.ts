import { createMacro, MacroParams } from "babel-plugin-macros"

import type { VisitNodeObject } from "@babel/traverse"
import { Program } from "@babel/types"

import linguiPlugin from "./index"
import { JsMacroName, JsxMacroName } from "./constants"

function macro({ state, babel, config }: MacroParams) {
  if (!state.get("linguiProcessed")) {
    state.opts = config
    const plugin = linguiPlugin(babel)

    const { enter, exit } = plugin.visitor.Program as VisitNodeObject<
      any,
      Program
    >

    enter(state.file.path, state)
    state.file.path.traverse(plugin.visitor, state)
    exit(state.file.path, state)

    state.set("linguiProcessed", true)
  }

  return { keepImports: true }
}

;[
  JsMacroName.defineMessage,
  JsMacroName.msg,
  JsMacroName.t,
  JsMacroName.useLingui,
  JsMacroName.plural,
  JsMacroName.select,
  JsMacroName.selectOrdinal,

  JsxMacroName.Trans,
  JsxMacroName.Plural,
  JsxMacroName.Select,
  JsxMacroName.SelectOrdinal,
].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throw new Error(
        `The macro you imported from "@lingui/core/macro" or "@lingui/react/macro" is being executed outside the context of compilation with babel-plugin-macros. ` +
          `This indicates that you don't have the babel plugin "babel-plugin-macros" configured correctly. ` +
          `Please see the documentation for how to configure babel-plugin-macros properly: ` +
          "https://github.com/kentcdodds/babel-plugin-macros/blob/main/other/docs/user.md"
      )
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
}) as { isBabelMacro: true }
