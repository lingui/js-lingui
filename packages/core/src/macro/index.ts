import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJs } from "./macroJs"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export type LinguiMacroOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

export const jsMacroTags = new Set([
  "defineMessage",
  "arg",
  "t",
  "plural",
  "select",
  "selectOrdinal",
])

let config: LinguiConfigNormalized

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { i18nImportModule, i18nImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsNodes = new Set<NodePath>()
  let needsI18nImport = false

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsMacroTags.has(tagName)) {
      nodes.forEach((node) => {
        jsNodes.add(node.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsNodesArray = Array.from(jsNodes)

  jsNodesArray.filter(isRootPath(jsNodesArray)).forEach((path) => {
    const macro = new MacroJs(babel, { i18nImportName, stripNonEssentialProps })
    if (macro.replacePath(path)) needsI18nImport = true
  })

  if (needsI18nImport) {
    addImport(
      babel.types,
      state.file.path.node,
      i18nImportModule,
      i18nImportName
    )
  }
}

;[...jsMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/core/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
