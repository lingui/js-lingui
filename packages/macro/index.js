module.exports = require("@lingui/babel-plugin-lingui-macro/macro")
