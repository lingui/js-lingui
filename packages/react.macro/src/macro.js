import { createMacro } from "babel-plugin-macros"
import { implementationPlugin as TransformReactPlugin } from "@lingui/babel-plugin-transform-react"

const importsToCarryOver = ["DateFormat", "NumberFormat"]

function macro({ references, state, babel }) {
  const { types: t } = babel
  const plugin = new TransformReactPlugin(babel)

  const toKeepImports = ["Trans"]

  // Trick the plugin into thinking we've processed an import
  plugin.setImportDeclarations(
    Object.keys(references).reduce((obj, key) => {
      if (key !== "default" || key !== "Trans") obj[key] = key
      return obj
    }, {})
  )

  for (let [tagName, tags] of Object.entries(references)) {
    if (importsToCarryOver.includes(tagName)) toKeepImports.push(tagName)

    tags.forEach(openingTag => {
      if (!t.isJSXOpeningElement(openingTag.container)) return // Exclude closing elements

      const node = openingTag.context.parentPath.container

      plugin.JSXElement({ node }, state.file)
    })
  }

  const linguiReactImport = state.file.path.node.body.find(
    importNode =>
      t.isImportDeclaration(importNode) &&
      importNode.source.value === "@lingui/react"
  )

  // Handle adding the import or altering the existing import
  if (linguiReactImport) {
    toKeepImports.forEach(name => {
      if (
        linguiReactImport.specifiers.findIndex(
          specifier => specifier.imported && specifier.imported.name === name
        ) === -1
      ) {
        linguiReactImport.specifiers.push(
          t.importSpecifier(t.identifier(name), t.identifier(name))
        )
      }
    })
  } else {
    state.file.path.node.body.unshift(
      t.importDeclaration(
        toKeepImports.map(name =>
          t.importSpecifier(t.identifier(name), t.identifier(name))
        ),
        t.stringLiteral("@lingui/react")
      )
    )
  }
}

export default createMacro(macro)
// For intellisense as these will be ignored by babel-macro
// Might want to copy PropTypes here?
export const Trans = () => {}
export const Plural = () => {}
export const Select = () => {}
export const SelectOrdinal = () => {}
export const DateFormat = () => {}
export const NumberFormat = () => {}
