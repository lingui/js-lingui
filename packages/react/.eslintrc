{
  "extends": ["../../.eslintrc", "plugin:react/recommended", "plugin:react-hooks/recommended"],
  "settings": {
    "react": {
      "version": "detect" // "detect" automatically picks the version of React that is installed
    }
  },
  "rules": {
    "react/prop-types": "off",
    "react/display-name": "off"
  }
}
