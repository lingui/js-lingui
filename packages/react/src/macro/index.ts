import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJSX } from "./macroJsx"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  LinguiMacroOpts,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export const jsxMacroTags = new Set([
  "Trans",
  "Plural",
  "Select",
  "SelectOrdinal",
])

let config: LinguiConfigNormalized

// todo, make memoization on the conf library level
function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { TransImportModule, TransImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsxNodes = new Set<NodePath>()

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsxMacroTags.has(tagName)) {
      // babel-plugin-macros return JSXIdentifier nodes.
      // Which is for every JSX element would be presented twice (opening / close)
      // Here we're taking JSXElement and dedupe it.
      nodes.forEach((node) => {
        // identifier.openingElement.jsxElement
        jsxNodes.add(node.parentPath.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsxNodesArray = Array.from(jsxNodes)

  jsxNodesArray.filter(isRootPath(jsxNodesArray)).forEach((path) => {
    const macro = new MacroJSX(babel, { stripNonEssentialProps })
    macro.replacePath(path)
  })

  if (jsxNodes.size) {
    addImport(
      babel.types,
      state.file.path.node,
      TransImportModule,
      TransImportName
    )
  }
}

;[...jsxMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/react/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
