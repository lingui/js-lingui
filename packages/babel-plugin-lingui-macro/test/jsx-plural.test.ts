import { macroTester } from "./macroTester"

macroTester({
  cases: [
    {
      code: `
        import { Plural } from '@lingui/react/macro';
        <Plural
          value={count}
          offset="1"
          _0="Zero items"
          few={\`\${count} items\`}
          other={<a href="/more">A lot of them</a>}
        />;
      `,
    },
    {
      name: "Plural macro could be renamed",
      code: `
        import { Plural as Plural2 } from '@lingui/react/macro';
        <Plural2
          value={count}
          one={"..."}
          other={"..."}
        />;
      `,
    },
    {
      name: "Should preserve reserved props: `comment`, `context`, `render`, `id`",
      code: `
        import { Plural } from '@lingui/react/macro';
        <Plural
          comment="Comment for translator"
          context="translation context"
          id="custom.id"
          render={() => {}}
          value={count}
          offset="1"
          _0="Zero items"
          few={\`\${count} items\`}
          other={<a href="/more">A lot of them</a>}
        />;
      `,
    },
    {
      code: `
        import { Trans, Plural } from '@lingui/react/macro';
        <Plural
          value={count}
          one={
            <Trans>
              <strong>#</strong> slot added
            </Trans>
          }
          other={
            <Trans>
              <strong>#</strong> slots added
            </Trans>
          }
        />;
      `,
    },
    {
      name: "Should return cases without leading or trailing spaces for nested Trans inside Plural",
      code: `
        import { Trans, Plural } from '@lingui/react/macro';
        <Plural
              one={
                <Trans>
                  One hello
                </Trans>
              }
              other={
                <Trans>
                  Other hello
                </Trans>
              }
              value={count}
        />;
      `,
    },
    {
      code: `
        import { Plural } from '@lingui/react/macro';
        <Plural
          id="msg.plural"
          render="strong"
          value={count}
          offset="1"
          _0="Zero items"
          few={\`\${count} items\`}
          other={<a href="/more">A lot of them</a>}
        />;
      `,
    },
    {
      code: `
        import { Trans, Plural } from '@lingui/react/macro';
        <Trans id="inner-id-removed">
          Looking for{" "}
          <Plural
            value={items.length}
            offset={1}
            _0="zero items"
            few={\`\${items.length} items \${42}\`}
            other={<a href="/more">a lot of them</a>}
          />
        </Trans>
      `,
    },
    {
      code: `
        import { Plural } from '@lingui/react/macro';
        <Plural
          value={count}
          _0="Zero items"
          one={oneText}
          other={<a href="/more">A lot of them</a>}
        />;
      `,
    },
    {
      filename: `jsx-plural-select-nested.js`,
    },
  ],
})
