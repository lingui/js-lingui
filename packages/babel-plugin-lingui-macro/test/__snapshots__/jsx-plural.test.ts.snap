// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#1 1`] = `
import { Plural } from "@lingui/react/macro";
<Plural
  value={count}
  offset="1"
  _0="Zero items"
  few={\`\${count} items\`}
  other={<a href="/more">A lot of them</a>}
/>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "tYX0sm",
      message:
        "{count, plural, offset:1 =0 {Zero items} few {{count} items} other {<0>A lot of them</0>}}",
      values: {
        count: count,
      },
      components: {
        0: <a href="/more" />,
      },
    }
  }
/>;

`;

exports[`#4 1`] = `
import { Trans, Plural } from "@lingui/react/macro";
<Plural
  value={count}
  one={
    <Trans>
      <strong>#</strong> slot added
    </Trans>
  }
  other={
    <Trans>
      <strong>#</strong> slots added
    </Trans>
  }
/>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "X8eyr1",
      message:
        "{count, plural, one {<0>#</0> slot added} other {<1>#</1> slots added}}",
      values: {
        count: count,
      },
      components: {
        0: <strong />,
        1: <strong />,
      },
    }
  }
/>;

`;

exports[`#6 1`] = `
import { Plural } from "@lingui/react/macro";
<Plural
  id="msg.plural"
  render="strong"
  value={count}
  offset="1"
  _0="Zero items"
  few={\`\${count} items\`}
  other={<a href="/more">A lot of them</a>}
/>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  render="strong"
  {
    /*i18n*/
    ...{
      id: "msg.plural",
      message:
        "{count, plural, offset:1 =0 {Zero items} few {{count} items} other {<0>A lot of them</0>}}",
      values: {
        count: count,
      },
      components: {
        0: <a href="/more" />,
      },
    }
  }
/>;

`;

exports[`#7 1`] = `
import { Trans, Plural } from "@lingui/react/macro";
<Trans id="inner-id-removed">
  Looking for{" "}
  <Plural
    value={items.length}
    offset={1}
    _0="zero items"
    few={\`\${items.length} items \${42}\`}
    other={<a href="/more">a lot of them</a>}
  />
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "inner-id-removed",
      message:
        "Looking for {0, plural, offset:1 =0 {zero items} few {{1} items {2}} other {<0>a lot of them</0>}}",
      values: {
        0: items.length,
        1: items.length,
        2: 42,
      },
      components: {
        0: <a href="/more" />,
      },
    }
  }
/>;

`;

exports[`#8 1`] = `
import { Plural } from "@lingui/react/macro";
<Plural
  value={count}
  _0="Zero items"
  one={oneText}
  other={<a href="/more">A lot of them</a>}
/>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "EQvNfC",
      message:
        "{count, plural, =0 {Zero items} one {{oneText}} other {<0>A lot of them</0>}}",
      values: {
        count: count,
        oneText: oneText,
      },
      components: {
        0: <a href="/more" />,
      },
    }
  }
/>;

`;

exports[`Plural macro could be renamed 1`] = `
import { Plural as Plural2 } from "@lingui/react/macro";
<Plural2 value={count} one={"..."} other={"..."} />;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "EMgKyP",
      message: "{count, plural, one {...} other {...}}",
      values: {
        count: count,
      },
    }
  }
/>;

`;

exports[`Should preserve reserved props: \`comment\`, \`context\`, \`render\`, \`id\` 1`] = `
import { Plural } from "@lingui/react/macro";
<Plural
  comment="Comment for translator"
  context="translation context"
  id="custom.id"
  render={() => {}}
  value={count}
  offset="1"
  _0="Zero items"
  few={\`\${count} items\`}
  other={<a href="/more">A lot of them</a>}
/>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  render={() => {}}
  {
    /*i18n*/
    ...{
      id: "custom.id",
      message:
        "{count, plural, offset:1 =0 {Zero items} few {{count} items} other {<0>A lot of them</0>}}",
      comment: "Comment for translator",
      context: "translation context",
      values: {
        count: count,
      },
      components: {
        0: <a href="/more" />,
      },
    }
  }
/>;

`;

exports[`Should return cases without leading or trailing spaces for nested Trans inside Plural 1`] = `
import { Trans, Plural } from "@lingui/react/macro";
<Plural
  one={<Trans>One hello</Trans>}
  other={<Trans>Other hello</Trans>}
  value={count}
/>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "oukcm6",
      message: "{count, plural, one {One hello} other {Other hello}}",
      values: {
        count: count,
      },
    }
  }
/>;

`;
